### PR TITLE
[Core] Base python solver

### PR DESCRIPTION
**Pull request by philbucher**
This PR adds a baseclass for the python solvers to the core. This is the second step after having the `AnalysisStage` now
I used the functions that are currently used in the FluidDynamics and the StructuralMechanics Solver and put them to the base-class

This PR is to discuss the interface, I think/hope that it will not be so difficult this time since I am not really doing new things and only collected what exists already

After an agreement has been reached I will add documentation and finish integrating it in the Structural Solver

Having this base class will allow us to call all the functions in the interface in the stage without having to check if a solver actually implements them

Note that I didn't include `Solve` since after @RiccardoRossi s explanation we should call the functions separately
Also I did not include `ImportModelPart`, this one is splitted into `ReadModelPart` and `PrepareModelPartForSolver`
The first one is common for the solvers and will not have to be overwritten in most derived solvers.
The second is specific to the solvers and does what the name suggests
This separation is needed one one hand to have the reading implemented only once and also in case the Solver should not read the modelpart (used by the optimization guys)

### CODE REVIEW

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [inline kratos/python_scripts/python_solver.py]: We can agree that the strategies will called in this cases, and then we can call this methods on the strategy

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [inline kratos/python_scripts/python_solver.py]: We can return the current delta time

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [inline kratos/python_scripts/python_solver.py]: We  can return 2 at least

**Review by loumalouomega — CHANGES_REQUESTED** [review]

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: ok

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: ok, good suggestion, esp for the structure since it cannot compute Dt yet

File: kratos/python_scripts/python_solver.py
Comment (jcotela) [reply kratos/python_scripts/python_solver.py]: Why 2?

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: I intentionally left this empty since theoretically a solver can have no strategy

e.g. an FSI solver does not have a strategy itself

File: kratos/python_scripts/python_solver.py
Comment (jcotela) [reply kratos/python_scripts/python_solver.py]: Do we need this function on the external interface of the base class? I don't think it is used outside of the solver right now, and we could think of solvers that don't have it.

File: kratos/python_scripts/python_solver.py
Comment (jcotela) [inline kratos/python_scripts/python_solver.py]: One thing I was wondering is if we should "hard-code" labels like this. If we call it from a derived solver, is it more useful for the label to be "[PythonSolver]" or "[DynamicStructuralSolver]"? I don't have a strong opinion on this, but it is a discussion I think we should have at some point.

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: Also true
Will remove it, the user should call `AdvanceInTime`

File: kratos/python_scripts/python_solver.py
Comment (jcotela) [inline kratos/python_scripts/python_solver.py]: Again, does this need to be a public method of the solver? Or should it work internally? Also, the name is a bit strange... If we are "the solver", why should we prepare something "for the solver"?

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [reply kratos/python_scripts/python_solver.py]: :+1:

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [reply kratos/python_scripts/python_solver.py]: As a default, the minimum buffer size. Obviously the idea is to override in the solvers

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [reply kratos/python_scripts/python_solver.py]: I thing, I remember or something... That in my analysis I added a method to return the label, that is overrided in each solver, so the solver always knows the name of the solver

Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: It would be easy to add a GetSolverName function that every solver implements

Might be useful if several solvers are used, which is likely to be the case in the future 

I dont habe a strong opinion though

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [inline kratos/python_scripts/python_solver.py]: for a static solver, 1 is enough...

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py]: I think it is good it is a public method, since one may want to override it. an example is if you have a "strange" problem and you need to build submodelparts on your own

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [inline kratos/python_scripts/python_solver.py]: what if we call this ImportModelPart, exactly with the implementation and you propose...that is, also followed by a function "Prepare"

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py]: what do you mean that it cannot have a strategy inside? it may even have multiple as long as they are hidden in the api...

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [inline kratos/python_scripts/python_solver.py]: Just to say that i invented this ComputingModelPart idea...and i hate it :-(.

I don't have a better idea however

**Review by RiccardoRossi — COMMENTED** [review]
i am making minor comments.

another one is that
def validate_and_transfer_matching_settings

i guess should be done in c++ with its own tests etc

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: Fine for me

I used a different name because others (e.g. the FSI solvers) are still using `ImportModelPart` and I would have silently broken those

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: it can have 0 to n strategies inside, that why I left it empty

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: I think it should be public
For fluids this is the function that would replace the elements, whereas in structure we e.g. assign the materials

File: kratos/python_scripts/python_solver.py
Comment (philbucher) [reply kratos/python_scripts/python_solver.py]: Another point for hardcoding is that if a solver calls the functions of the baseclass (e.g. `Initialize`) then we would have multiple `Initializing done` prints with the same solver name which I think would be quite confusing

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py]: this argument makes sense.

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [reply kratos/python_scripts/python_solver.py]: sorry i misread what you wrote.

**Comment by philbucher** [comment]
I finished the base-class and the adaption of the Structural Solvers
=> ready to be properly reviewed

I ran the tests for the involved applications

@jcotela I tried to do it also for the Fluid but there is a bit of refactoring needed, so I would prefer if it would be done in a separate PR

**Comment by jcotela** [comment]
@philbucher I'll make a pass of refactoring for fluid solvers once this is merged.

**Comment by philbucher** [comment]
ping @KratosMultiphysics/technical-committee

File: kratos/python_scripts/python_solver.py
Comment (RiccardoRossi) [inline kratos/python_scripts/python_solver.py]: current time can be obtained as 

      model_part.ProcessInfo[TIME]

why do you need to pass it?

**Comment by philbucher** [comment]
ping @KratosMultiphysics/technical-committee I added the [wiki](https://github.com/KratosMultiphysics/Kratos/wiki/Common-Python-Interface-of-Applications-for-Users) for the AnalysisStage and the PythonSolver

Please have a look such that we can proceed with this and #2135

File: kratos/python_scripts/python_solver.py
Comment (pooyan-dadvand) [inline kratos/python_scripts/python_solver.py:110]: I would add some trivial checks (like settings is a Parameter, input file exists, etc.)

Comment (philbucher) [reply kratos/python_scripts/python_solver.py:110]: the type of the input is checked in the constructor 

I thought of this function to be called before the SolutionLoop (I will add it to the AnalysisStage)
=> It should call methods like Strategy.Check() to validate if the input is correct

Therefore it cannot validate if the input file has been found
(btw for this one the error is quite clear already:
`RuntimeError: Error: Error opening mdpa file : Example.mdpa`)

**Comment by philbucher** [comment]
I did the modifications as requested by @RiccardoRossi  => passing the `Model` instead of the `ModelPart` to the solver

@KratosMultiphysics/technical-committee please have a look since this has to be merged before #2135 

After this is merged I will solve the "conflicts" in #2135 to finalize everything

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
@philbucher feel free to merge both #2135 and this PR

**Review by RiccardoRossi — DISMISSED** [review]
i am approving. @philbucher wait if someone blocks it tomorrow in the morning and otherwise merge...

**Comment by loumalouomega** [comment]
@philbucher I have been doing experiments with a branch, merging this PR and #2135, and for example the fluid does not work anymore, @jcotela  do you mind if I repair that?

File: kratos/python_scripts/python_solver.py
Comment (loumalouomega) [inline kratos/python_scripts/python_solver.py]: There are two of these, there is the ImportModelPart and the _ImportModelPart, why?

Comment (loumalouomega) [reply kratos/python_scripts/python_solver.py]: OK, deprecated, I see

**Comment by philbucher** [comment]
@loumalouomega please wait with pushing to this Branch, I am fixing FSI and your branch before merging this

@jcotela and I know abt the problems, the fluid will be done in a separate PR before #2135 

Please give me until this evening then everything should be working

**Comment by loumalouomega** [comment]
Don't worry, I am working locally, but right no I am havig problems to make work the model properly with the main model part

**Comment by philbucher** [comment]
ok, but be aware that I am also working on this

I wil push the changes regarding the solver to this branch in ~2hrs and afterwards will integrate the changes in the AnalyisStage-Pr

**Comment by philbucher** [comment]
this is the final set of modifications
@loumalouomega @rubenzorrilla I updated your apps, the tests that run on master are still working

@KratosMultiphysics/technical-committee there were some minor things I still had to fix, in relation with #2135 , but I didn't change anything major

I will merge this branch now to #2135 to fully apply the changes there

@RiccardoRossi switching to the Model (=> #2137 ) should now only be a minor change

**Review by loumalouomega — APPROVED** [review]
OK

**Comment by philbucher** [comment]
I will merge this tmr morning if none blocks/complains

**Comment by rubenzorrilla** [comment]
Apart of the fact that I did not have time to check it until now, this breaks retro compatibility (at least for the FSI app). I guess that this is something we have to live with.

In any case, it is not compatible with the current Main script generated by the interface.

**Pull request merged**
